src/p2p/net_node.inl

```diff
@@ -1204,48 +1285,55 @@ namespace nodetool
 
     if (!m_exclusive_peers.empty()) return true;
 
-    size_t start_conn_count = get_outgoing_connections_count();
-    if(!m_peerlist.get_white_peers_count() && m_seed_nodes.size())
+    // Only have seeds in the public zone right now.
+
+    size_t start_conn_count = get_public_outgoing_connections_count();
+    if(!get_public_white_peers_count() && m_seed_nodes.size())
     {
       if (!connect_to_seed())
         return false;
     }
 
     if (!connect_to_peerlist(m_priority_peers)) return false;
 
-    size_t expected_white_connections = (m_config.m_net_config.max_out_connection_count*P2P_DEFAULT_WHITELIST_CONNECTIONS_PERCENT)/100;
-
-    size_t conn_count = get_outgoing_connections_count();
-    if(conn_count < m_config.m_net_config.max_out_connection_count)
+    for(auto& zone : m_network_zones)
     {
-      if(conn_count < expected_white_connections)
-      {
-        //start from anchor list
-        if(!make_expected_connections_count(anchor, P2P_DEFAULT_ANCHOR_CONNECTIONS_COUNT))
-          return false;
-        //then do white list
-        if(!make_expected_connections_count(white, expected_white_connections))
-          return false;
-        //then do grey list
-        if(!make_expected_connections_count(gray, m_config.m_net_config.max_out_connection_count))
-          return false;
-      }else
+      size_t expected_white_connections = (zone.second.m_config.m_net_config.max_out_connection_count*P2P_DEFAULT_WHITELIST_CONNECTIONS_PERCENT)/100;
+
+      size_t conn_count = get_outgoing_connections_count(zone.second);
+      if(conn_count < zone.second.m_config.m_net_config.max_out_connection_count)
       {
-        //start from grey list
-        if(!make_expected_connections_count(gray, m_config.m_net_config.max_out_connection_count))
-          return false;
-        //and then do white list
-        if(!make_expected_connections_count(white, m_config.m_net_config.max_out_connection_count))
-          return false;
+        if(conn_count < expected_white_connections)
+        {
+          //start from anchor list
+          if(!make_expected_connections_count(zone.second, anchor, P2P_DEFAULT_ANCHOR_CONNECTIONS_COUNT))
+            return false;
+          //then do white list
+          if(!make_expected_connections_count(zone.second, white, expected_white_connections))
+            return false;
+          //then do grey list
+          if(!make_expected_connections_count(zone.second, gray, zone.second.m_config.m_net_config.max_out_connection_count))
+            return false;
+        }else
+        {
+          //start from grey list
+          if(!make_expected_connections_count(zone.second, gray, zone.second.m_config.m_net_config.max_out_connection_count))
+            return false;
+          //and then do white list
+          if(!make_expected_connections_count(zone.second, white, zone.second.m_config.m_net_config.max_out_connection_count))
+            return false;
+        }
```

Comment:
That doesn't seem right.  You probably want to do this (the existing code) for the public zone, then have tor/other try one or two extra peers. If you're unlucky, trying all zones might get you most of your peers on tor, where you can't sync.

Reply:
Actually, I think it's not OK. It means we'll always (often) connect to the same tor service, and thus be a lot easier to get our privacy broken.

Reply:
I don't understand why the behavior would be different for Tor than IP. There is a peerlist per zone too - only the serialized file is combined. I might need to go back and review this closer.

Reply:
Anchor peers are fixed over several runs of the daemon. If you use those for Tor, it means your node will always connect to the same Tor hidden services. This might make it easier to track someone over time since you have extra info at the hidden service's end. Maybe this is negligible given Tor, but using non-anchor peers here removes that possibility.

Reply:
Which extra information will the hidden service have about connecting nodes? Timing, height, first-seen transactions, inventory of transactions seen, and connected nodes? I wonder because tor's approach to entry nodes is that perhaps counter-intuitively, but based on research, connecting to the same entry tor node each time lowers your probability of profiling. I know there is a difference connecting to tor circuit nodes and monero p2p network nodes, but I wonder if a similar effect would apply or not?

https://www.torproject.org/docs/faq.html.en#EntryGuards

Reply:
In this case, the connection would be to the same hidden service. Tor will decide about the entry guard. But good point, maybe it is not actually a problem, and even a good thing...

Reply:
I left the anchor behavior untouched for now. I'm not sure of the best approach, but this can be tweaked separately later.